### PR TITLE
Optimize equality comparison for small str with fixed size

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: ref #5294

After #5545

### What is changed and how it works?

- optimize equality comparison for small str with fixed size ( size <= 16) like `select ... from ... where xxx = 'xxx' ...`
- add function `LoopOneColumnCmpEqFixedStr<>` to for such equality comparison
- add function `memcmp_eq_fixed_size<>`.

----

```
FLATTEN_INLINE_PURE
inline bool memcmp_eq3(const char * a, const char * b)
{
    return memcmp_eq2(a, b) & memcmp_eq2(a + 1, b + 1);
}
```

If use `&&` and cpu failed to predict `cmp -> jne`, the pipeline will be broken.
```
    movzx   eax, word ptr [rdi]
    cmp     ax, word ptr [rsi]
    jne     ...
    movzx   eax, word ptr [rdi + 1]
    cmp     ax, word ptr [rsi + 1]
    sete    al
    ret
```
Use `&` to reduce unnecessary branch. Instructions like (1) and (2) are independent(same for (3) and (4)), it's friendly for parallelism.
```
    movzx   eax, word ptr [rdi]         // (1)
    movzx   ecx, word ptr [rdi + 1]     // (2)
    xor     ax, word ptr [rsi]          // (3)
    xor     cx, word ptr [rsi + 1]      // (4)
    or      cx, ax
    sete    al
    ret
```

### Benchmark

ENV
- tpch-100
- 1 tiflash
- limit cpu up to 200%
- original commit: 30fc64c323d262e0faad10150d31025d06efe517

---

SQL

```
select count(1) from lineitem where L_SHIPMODE = 'zzzz';
```

Time(s) | Original | Optimized |   |   | Improvement
-- | -- | -- | -- | -- | --
  | 9.15 | 7.52 |   |   |  
  | 9.33 | 7.62 |   |   |  
  | 9.12 | 7.58 |   |   |  
  | 9.23 | 7.57 |   |   |  
  | 9.14 | 7.65 |   |   | AVG(Original) / AVG(Optimized) - 1.0
AVG | 9.194 | 7.588 |   | Optimized : Original | 21.16%

---

SQL

- ref [tpch-q10](https://github.com/pingcap/go-tpc/blob/69c6de2bd5157131427f6a31e083a1767a59002f/tpch/query.go#L294)

```
select count(1) from lineitem where L_RETURNFLAG = 'R';
```

```
MySQL [tpch100]> select count(1) from lineitem where L_RETURNFLAG = 'R';
+-----------+
| count(1)  |
+-----------+
| 148067261 |
+-----------+
```

Time(s) | Original | Optimized |   |   | Improvement
-- | -- | -- | -- | -- | --
  | 12.85 | 8.56 |   |   |  
  | 12.87 | 8.64 |   |   |  
  | 12.86 | 8.45 |   |   |  
  | 12.75 | 8.51 |   |   |  
  | 12.76 | 8.64 |   |   | AVG(Original) / AVG(Optimized) - 1.0
AVG | 12.818 | 8.56 |   | Optimized : Original | 49.74%

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: libs/libcommon/include/common/fixed_mem_eq.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+{
+    const uint16_t ax = *reinterpret_cast<const uint16_t *>(a);
+    const uint16_t bx = *reinterpret_cast<const uint16_t *>(b);
+    return ax == bx;
+}
+
+FLATTEN_INLINE_PURE
+inline bool memcmp_eq3(const char * a, const char * b)
+{
+    return memcmp_eq2(a, b) & memcmp_eq2(a + 1, b + 1);
```
Comment (windtalker) [inline libs/libcommon/include/common/fixed_mem_eq.h:48]: 1. why use `&` instead of `&&`
2. why not use `memcmp_eq2(a,b) && memcmp_eq1(a+2,b+2)`?

Comment (solotzg) [reply libs/libcommon/include/common/fixed_mem_eq.h:48]: To reduce branch. If cpu failed to predict `cmp and jne`, the pipeline will be cleared and restarted. 

----

use `&&`
```
        movzx   eax, word ptr [rdi]
        cmp     ax, word ptr [rsi]
        jne     .LBB6_1
        movzx   eax, word ptr [rdi + 1]
        cmp     ax, word ptr [rsi + 1]
        sete    al
        ret
```

----

use `&`
```
        movzx   eax, word ptr [rdi]  // (1)
        movzx   ecx, word ptr [rdi + 1]  // (2)
        xor     ax, word ptr [rsi]  // (3)
        xor     cx, word ptr [rsi + 1]  // (4)
        or      cx, ax
        sete    al
        ret
```
Instructions like (1) and (2) are independent(same as (3) and (4)), can be executed parallelly.

----

**Review by windtalker — APPROVED** [review]
LGTM

File: libs/libcommon/include/common/fixed_mem_eq.h [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        return memcmp_eq##s(a, b); \
+    }
+#endif
+
+    static_assert(k >= 0);
+    static_assert(k <= 32);
+
+    if constexpr (k > 16)
+    {
+        return memcmp_eq_fixed_size<16>(a, b) & memcmp_eq_fixed_size<k - 16>(a + 16, b + 16);
```
Comment (zanmato1984) [inline libs/libcommon/include/common/fixed_mem_eq.h:85]: Any consideration about using bitwise and (`&`) instead of logical and (`&&`)?

Comment (zanmato1984) [reply libs/libcommon/include/common/fixed_mem_eq.h:85]: Just saw the other comment address this question.

But maybe we can add some comments here.

**Review by zanmato1984 — APPROVED** [review]
Otherwise LGTM.

File: libs/libcommon/include/common/fixed_mem_eq.h (hunk 2 above)
Comment (solotzg) [reply libs/libcommon/include/common/fixed_mem_eq.h:85]: Same as https://github.com/pingcap/tiflash/pull/5569#discussion_r945431409

**Pull request merged**